Adds support to connections with Thrift over HTTP transport.

joaopedroantonio opened the pull request:
- Supports HTTP transport for Thrift protocol.
- Three types of authentication supported: NONE, NOSASL, BASIC and KERBEROS.
- BASIC authentication is useful when the Thrift HTTP interface is behind a proxy (e.g. in Azure HDInsight clusters).

Commit:
Adds support to connections with Thrift over HTTP transport.

- Supports HTTP transport for Thrift protocol.
- Three types of authentication supported: NONE, BASIC and KERBEROS.

joaopedroantonio commented:
This PR solves issue https://github.com/dropbox/PyHive/issues/69. Let me know what you think and if it looks good for you I'll invest time in the tests that are missing in this PR. :)

bkyryliuk commented:
will take a pass on it next week, unfortunately can't get to this earlier

joaopedroantonio commented on pyhive/hive.py:
```diff
@@ -94,12 +96,24 @@ def connect(*args, **kwargs):
+# TODO
+# Setting the Cookie in the headers should be implemented in the thrift library.
+# We'll keep this here until that change is available in there.
+class TCookieHttpClient(thrift.transport.THttpClient.THttpClient):
```
In the meantime, this commit was merged into Thrift master branch, we just have to wait for a new Thrift release to get rid of this TCookieHttpClient:

https://github.com/apache/thrift/commit/69642f389a06f5ba1b374de52c6b0e29892035d8

gthomas-slack replied:
It looks like this fix has been included in the 0.13.0 Release

bkyryliuk commented:
@joaopedroantonio I'm aligned with the approach here, let's add the unit tests & add the test plan in the PR description if you don't mind.

joaopedroantonio commented:
Thank you for the feedback @bkyryliuk, will start working on the tests between this weekend and start of next week.

Commit:
Implement unit tests.

MinuMichaelOlassayil commented:
Any update on this?

joaopedroantonio commented:
> Any update on this?

Oh well, tbh this got lost on my backlog, thanks for the bump. :) Will try to wrap it up this week.

joaopedroantonio commented:
Added unit tests for different HTTP scenarios and fixed a few issues in the process.

@bkyryliuk the codecov checks failed but I guess that's mostly because I moved the binary transport implementation and there are a few line/branch misses. Do you think it's relevant to cover those misses or can we move forward as it is?

joaopedroantonio committed:
Add unit tests to increase code coverage.

joaopedroantonio commented:
@bkyryliuk, any idea when you might be able to pick this up? :)

joaopedroantonio commented:
Hello again everyone! Can you give some feedback on this PR? Thanks! (cc @bkyryliuk )

gthomas-slack commented on pyhive/hive.py:
```diff
@@ -207,6 +196,109 @@ def sasl_factory():
+        if http_path is None:
+            http_path = '/'
+
+        socket = TCookieHttpClient('http://{}:{}{}'.format(host, port, http_path))
```
It would be great if we could also pass through a `http_protocol`, either `http` or `https` and have this set.

gthomas-slack commented:
I've tested this PR and this is working nicely for our HiveServer2 service running over `http`. Awesome work @joaopedroantonio - would love to see this get merged

gthomas-slack commented:
For our particular use-case we need a few more features:
* Support for connecting to a `https` endpoint
* Support for passing through custom http headers
* Support for passing through custom cookie values

I don't think these should block this PR though, I'd be happy to attempt to add these features once this initial version is in, unless of course you want to add them :)

gthomas-slack commented on pyhive/hive.py:
```diff
@@ -207,6 +196,109 @@ def sasl_factory():
+        if auth is None:
+            auth = 'NONE'
+        if http_path is None:
+            http_path = '/'
```
It could be better to default this to `/cliservice`, as that seems to be the default path HiveServer2 provides

joaopedroantonio commented:
@gthomas-slack sorry for the late reply! I'm glad this was helpful! I'll take a better look at your comments over the weekend and I'm happy to include your suggestions if they aren't too complex. :)

lvazquez commented:
For thrift=0.13.0 (current version in conda/pip) it doesn't work  with the included **TCookieHttpClient** wrapper class, and I get the following error in the connect constructor when executing the first "use default" query:
```
  File "/root/superset/pyhive/pyhive/hive.py", line 189, in __init__
    cursor.execute('USE `{}`'.format(database))
  File "/root/superset/pyhive/pyhive/hive.py", line 459, in execute
    response = self._connection.client.ExecuteStatement(req)
  File "/root/superset/pyhive/TCLIService/TCLIService.py", line 280, in ExecuteStatement
    return self.recv_ExecuteStatement()
  File "/root/superset/pyhive/TCLIService/TCLIService.py", line 292, in recv_ExecuteStatement
    (fname, mtype, rseqid) = iprot.readMessageBegin()
  File "/root/superset/pyhive/env/lib/python3.7/site-packages/thrift/protocol/TBinaryProtocol.py", line 148, in readMessageBegin
    name = self.trans.readAll(sz)
  File "/root/superset/pyhive/env/lib/python3.7/site-packages/thrift/transport/TTransport.py", line 68, in readAll
    raise EOFError()
```
After replacing TCookieHttpClient with the original THttpClient it seems to work ok for http connections (no SSL support included).

lvazquez commented:
I believe the parameters name for the HTTP Thrift should be renamed to match the (simpler) names used in the original JDBC Thrift URI. Also 'binary', 'http', 'sasl' are not named "protocols" but transport "modes" according to JDBC/Hive documentation
[Connection URL When HiveServer2 Is Running in HTTP Mode](https://cwiki.apache.org/confluence/display/Hive/HiveServer2+Clients#HiveServer2Clients-ConnectionURLWhenHiveServer2IsRunninginHTTPMode)

Example Thrift JDBC over HTTPS URI:
`jdbc:hive2://{host}:{port}/{database};transportMode=http;httpPath=/hive;ssl=true;sslTrustStore=truststore.jks;trustStorePassword=changeit`

Based on this it seems the names should be:
*` thrift_transport_protocol` -> `transport_mode` (transportMode in JDBC/Thrift)
* `http_path` -> `http_path `(httpPath in JDBC/Thrift)

The remaining parameters needed to support HTTPS (HTTP over SSL) should be:
* ssl = True/False (could also be handled with `transport_mode=https` but this seems cleaner).
* `cacert`, ` ca_cert ` or `cafile` (Python uses plain PEM cert files instead of Java keystores/truststores
* OPTIONAL: `cert_file`, `key_file` and `key_pass` - to support SSL client authentication (as intended in #306)

kent-pawar commented:
Hi @joaopedroantonio . Bumping this. Thank you for working on this! Could you pls look at the new feedback. Thank you

mhconradt commented:
I've tested this with my Spark Thrift HiveServer2 deployment behind a load balancer on Kubernetes and it works great.
I'm looking forward to using this to connect Superset to my Spark Thrift server and would like to see this accepted soon.
Thanks @joaopedroantonio and everyone in the thread.

gthomas-slack commented:
For anyone who needs to use http/https HiveServer2 in production right now, It's actually possible with the current release of `PyHive`, as the HiveServer2 Connection class exposes a `thrift_transport` which allows you to pass in a custom configured transport to use: https://github.com/dropbox/PyHive/blob/master/pyhive/hive.py#L112-L113

Here is a basic example using HTTPS and adding a custom authentication header:

```python
from pyhive import hive
from os import getenv
import base64

import thrift.transport.THttpClient


def thrift_http_transport():
    transport = thrift.transport.THttpClient.THttpClient(uri_or_host='https://my-hiveserver2.com:443/cliservice')

    auth_credentials = '{}:{}'.format('test', 'test').encode('UTF-8')
    auth_credentials_base64 = base64.standard_b64encode(auth_credentials).decode('UTF-8')
    transport.setCustomHeaders(
        {
            'Authorization': 'Basic {}'.format(auth_credentials_base64),  # HiveServer2 BASIC auth
            'X-Auth-Token': 'xxx' # Custom header to auth with some kind of middleware
        }
    )
    return transport


conn = hive.connect(thrift_transport=thrift_http_transport())
cursor = conn.cursor()
cursor.execute("""SELECT SUM(1) from model.dim_date""")
data = cursor.fetchall()
cursor.close()
conn.close()
print(data)

```

mithunjmistry commented:
Can we use http protocol with SQLAlchemy?

gthomas-slack commented:
FYI there is currently a bug using HTTP transport connections when using thrift `0.15.0`, some details here: 
https://github.com/dropbox/PyHive/issues/417#issuecomment-961257993

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/dropbox/PyHive?pullRequest=325) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/dropbox/PyHive?pullRequest=325) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: Joao Antonio<br/>:x: joaopedroantonio<br/><hr/>**Joao Antonio** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/dropbox/PyHive?pullRequest=325) it.</sub>

realtonyyoung commented:
Hi, can anyone respond on whether this has been merged or not?SQLAlchemy is failing with http transport and really need a fix
